src/test/kotlin/org/rust/lang/core/type/RsImplicitTraitsTest.kt

```diff
@@ -0,0 +1,165 @@
+import org.intellij.lang.annotations.Language
+import org.rust.lang.core.psi.RsTraitItem
+import org.rust.lang.core.psi.RsTypeReference
+import org.rust.lang.core.resolve.ImplLookup
+import org.rust.lang.core.types.ty.TyBool
+import org.rust.lang.core.types.ty.TyChar
+import org.rust.lang.core.types.ty.TyFloat
+import org.rust.lang.core.types.ty.TyInteger
+import org.rust.lang.core.types.type
+
+class RsImplicitTraitsTest : RsTypificationTestBase() {
+
+    fun `test primitive types are Sized`() = checkPrimitiveTypes("Sized")
+
+    fun `test array is Sized`() = doTest("""
+        fn foo() -> [i32; 2] { unimplemented!() }
+                  //^ Sized
+    """)
+
+    fun `test slice is not Sized`() = doTest("""
+        fn foo() -> Box<[i32]> { unimplemented!() }
+                      //^ !Sized
+    """)
+
+    fun `test trait object is not Sized`() = doTest("""
+        trait Foo {}
+        fn foo() -> Box<Foo> { unimplemented!() }
+                       //^ !Sized
+    """)
+
+    fun `test enum is Sized`() = doTest("""
+        enum FooBar { Foo, Bar }
+        fn foo() -> FooBar { unimplemented!() }
+                      //^ Sized
+    """)
+
+    fun `test struct is Sized`() = doTest("""
+        struct Foo { foo: i32 }
+        fn foo() -> Foo { unimplemented!() }
+                    //^ Sized
+    """)
+
+    fun `test struct with DST field is not Sized`() = doTest("""
+        struct Foo { foo: i32, bar: [i32] }
+        fn foo() -> Box<Foo> { unimplemented!() }
+                       //^ !Sized
+    """)
+
+    fun `test tuple struct is Sized`() = doTest("""
+        struct Foo(i32);
+        fn foo() -> Foo { unimplemented!() }
+                    //^ Sized
+    """)
+
+    fun `test tuple struct with DST field is not Sized`() = doTest("""
+        struct Foo(i32, [i32]);
+        fn foo() -> Box<Foo> { unimplemented!() }
+                       //^ !Sized
+    """)
+
+    fun `test empty struct is Sized`() = doTest("""
+        struct Foo;
+        fn foo() -> Foo { unimplemented!() }
+                    //^ Sized
+    """)
+
+    fun `test tuple is Sized`() = doTest("""
+        fn foo() -> (i32, bool) { unimplemented!() }
+                  //^ Sized
+    """)
+
+    fun `test tuple with DST field is not Sized`() = doTest("""
+        fn foo() -> Box<(i32, [i32])> { unimplemented!() }
+                      //^ !Sized
+    """)
+
+    fun `test reference is Sized`() = doTest("""
+        fn foo() -> &i32 { unimplemented!() }
+                  //^ Sized
+    """)
+
+    fun `test pointer is Sized`() = doTest("""
+        fn foo() -> *const u32 { unimplemented!() }
+                   //^ Sized
+    """)
+
+    fun `test type parameter is Sized by default`() = doTest("""
+        fn foo<T>() -> T { unimplemented!() }
+                     //^ Sized
+    """)
+
+    fun `test type parameter with ?Sized bound is not Sized`() = doTest("""
+        fn foo<T: ?Sized>() -> Box<T> { unimplemented!() }
```

Comment:
Let's add a test where `T: Sized` explicitly: `fn foo<T: Sized>`. It allows to make sure that we handle `?` correctly

Reply:
Done